ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/diff/StateTraceGenerator.java

```diff
@@ -85,129 +84,133 @@ private Stream<StateDiffTrace> generate(
-      stateDiff.put(accountAddress.toHexString(), accountDiff);
+      // In pre-state mode, every touched account is included.
+      stateDiffResult.put(address.toHexString(), accountDiff);
     }
   }
 
   /**
-   * Calculate the storage differences for an account between the pre-transaction and
-   * post-transaction world states.
-   *
-   * <p>Behavior differs based on the mode:
-   *
-   * <ul>
-   *   <li><b>Diff mode</b> (isPreState = false): Only includes storage slots whose values changed
-   *       compared to the pre-transaction state. New accounts will only include non-zero storage
-   *       slots.
-   *   <li><b>Pre-state mode</b> (isPreState = true): Includes all storage slots that were touched,
-   *       even if the value did not change. New accounts will include both zero and non-zero
-   *       values.
-   * </ul>
+   * Get the {@link WorldUpdater} at the transaction scope.
    *
-   * @param updatedAccount The account state after transaction execution (mutable).
-   * @param rootAccount The account state before the transaction, or {@code null} if the account was
-   *     newly created during the transaction.
-   * @param isPreState Whether pre-state mode is active.
-   * @return A map of storage slot key → {@link DiffNode}, representing the diff of each slot.
+   * <p>Traverses up the world updater chain to retrieve the correct layer.
    */
-  private Map<String, DiffNode> calculateStorageDiff(
-      final MutableAccount updatedAccount, final Account rootAccount, final boolean isPreState) {
-
-    final Map<String, DiffNode> storageDiff = new TreeMap<>();
-    // Iterate through all updated storage slots in this transaction
-    updatedAccount
-        .getUpdatedStorage()
-        .forEach(
-            (key, newValue) -> {
-              if (rootAccount == null) {
-                // Case 1: Account did not exist before this transaction
-                // Record non-zero slots in diff mode, or all slots in pre-state mode
-                if (!UInt256.ZERO.equals(newValue) || isPreState) {
-                  storageDiff.put(key.toHexString(), new DiffNode(null, newValue.toHexString()));
-                }
-              } else {
-                // Case 2: Account existed pre-transaction
-                // Record differences, or everything if pre-state mode is enabled
-                if (!rootAccount.getStorageValue(key).equals(newValue) || isPreState) {
-                  final String originalValueHex = rootAccount.getStorageValue(key).toHexString();
-                  storageDiff.put(
-                      key.toHexString(), new DiffNode(originalValueHex, newValue.toHexString()));
-                }
-              }
-            });
-    return storageDiff;
+  private WorldUpdater getTransactionUpdater(final List<TraceFrame> traceFrames) {
+    // Transaction updater sits two levels above the frame's updater
+    return traceFrames.getFirst().getWorldUpdater().parentUpdater().get().parentUpdater().get();
```

Comment:
seems to be dangerous, if someone change the updater logic it will impact this code . maybe not possible but we cannot find another way to have it or pass it ?

Reply:
Any ideas? This comes from the original code